### PR TITLE
NodeMaterial revision

### PR DESCRIPTION
**Pull request by sunag**
This includes fixes and example of @DanielSturk, irradiance feature and better interface for Color Space using nodes.

Related:
https://github.com/mrdoob/three.js/pull/16788
https://github.com/mrdoob/three.js/pull/16668
https://github.com/mrdoob/three.js/issues/16375

## Changes
- unify `Math1Node`, `Math2Node`, `Math3Node` to `MathNode`
- new `ColorSpaceNode`
- new example `webgl_materials_envmaps_hdr_nodes.html`
- rename `Node.parse` to `Node.analyze`
- rename `Node.buildCode` to `Node.flow`
- rename `Node.eval` to `Node.parse`
- move `NodeBuilder.optimize = false` to `NodeBuilder.context.ignoreCache = true`
- fix irradiance `LDR`, `HDR` and `RGBM16`
- fix `RectAreaLight`

https://rawgit.com/sunag/three.js/dev-fix-TextureCubeNode/examples/index.html?q=nodes#webgl_materials_envmaps_hdr_nodes

image

### CODE REVIEW

**Comment by sunag** [comment]
How should we call the `vector2`?
This is the indirect(irradiance) cube coords vector.
/ping @WestLangley @bhouston @mrdoob 

Select `vector2` in `scope` option:
https://rawgit.com/sunag/three.js/dev-fix-TextureCubeNode/examples/webgl_materials_nodes.html?e=node-reflect

**Comment by sunag** [comment]
Here the original code:
https://github.com/mrdoob/three.js/blob/2ad018c7c7ad103c9da9d7f22b3154537384ab26/src/renderers/shaders/ShaderChunk/envmap_physical_pars_fragment.glsl.js#L30

**Comment by sunag** [comment]
Opts it is a `world normal` with `flipEnvMap`!?

```javascript
new THREE.NormalNode( THREE.NormalNode.WORLD )
```

**Comment by DanielSturk** [comment]
Great progress! Looking forward to seeing these changes made

**Comment by sunag** [comment]
Thanks @DanielSturk! 

Done the fixes and update the example. The visual result should be the same for both now`MeshStandardNodeMaterial` and `MeshStandardMaterial`.

https://rawgit.com/sunag/three.js/dev-fix-TextureCubeNode/examples/webgl_materials_envmaps_hdr_nodes.html

**Comment by sunag** [comment]
@donmccurdy following your tip.
https://github.com/mrdoob/three.js/pull/16796/commits/4665931e7b3a5af40066b015dfc2c42ab1c12bcb

File: examples/js/nodes/materials/NodeMaterial.js [hunk 1]
```diff
@@ -18,6 +18,22 @@ function NodeMaterial( vertex, fragment ) {
 
 	this.updaters = [];
 
+	// onBeforeCompile can't be in the prototype because onBeforeCompile.toString varies per material
```
Comment (donmccurdy) [inline examples/js/nodes/materials/NodeMaterial.js]: Where does `onBeforeCompile.toString` get called?

**Comment by donmccurdy** [comment]
@sunag thank you! :)

***

In words, how would you define 'analyze', 'parse', and 'flow' as they relate to NodeMaterial? I'm having some trouble understanding what "analyze" means — the word usually doesn't imply side effects, but in this case `analyze()` calls methods like `build()` and `addFlow()`.

**Comment by sunag** [comment]
### Primary process

1. **analyze()** It is necessary to analyze and validate all nodes and slots(channels), it is will generate info necessary for build the code, share nodes(create temp vars), optimization as well the variable type and the return value.

2. **flow()** It generates codes, includes functions, defines, uniforms, and other codes need to work the shader. It depends on `analyze` be execute before `flow` to generate info necessary to build the code/shader. `flow` returns all date of the slot/channel.

3. **analyzeAndFlow()** Call the two methods `analyze` and after the `flow`. It does not optimize the shader if used many slots/channels, for the reason, it is mostly used in `transform` slot of the `vertex shader`.

### Secundary process

4. **build** returns the final value, generate by `generate` function or the variable generate by `NodeTemp` just like any other optimization node.

5. **generate** create the snippets fundamental to generate the code. Here too, is responsible to include functions, defines, etc. Return only the final value converted to output using `format` function.

The process superficially is similar to this diagram:

Compiler

**Comment by sunag** [comment]
`parse` is used only to parse GLSL code, example is FunctionNode and StructNode

File: examples/js/nodes/materials/NodeMaterial.js (hunk 1 above)
Comment (sunag) [reply examples/js/nodes/materials/NodeMaterial.js]: It is used to create a hash modification in onBeforeCompile. More details:
https://github.com/mrdoob/three.js/pull/14442
https://github.com/mrdoob/three.js/pull/16668

Comment (sunag) [reply examples/js/nodes/materials/NodeMaterial.js]: https://github.com/mrdoob/three.js/pull/15674

Comment (donmccurdy) [reply examples/js/nodes/materials/NodeMaterial.js]: I don't see anything calling `toString()` in these PRs?

**Comment by donmccurdy** [comment]
Thanks! I'm confused, but I appreciate the detailed writeup. Probably I need to read the code more and see what calls what. 😅

File: examples/js/nodes/materials/NodeMaterial.js (hunk 1 above)
Comment (mvc1006) [reply examples/js/nodes/materials/NodeMaterial.js]: I'm confused,threejs is a new unfamiliar technology for me ,how to learning it ?

Comment (sunag) [reply examples/js/nodes/materials/NodeMaterial.js]: In 15674 have a commit calling toString
https://github.com/mrdoob/three.js/pull/15674/commits/07236bb765fd2d2231196ca89125b335144241e1
The description of the problem:
https://github.com/mrdoob/three.js/pull/15674#issuecomment-459417579

I think that this approach will be improved in the future...
https://github.com/mrdoob/three.js/blob/7fe9a41cb6930d8600305e8ee85830b011e461bb/src/renderers/webgl/WebGLPrograms.js#L249

**Comment by mrdoob** [comment]
@donmccurdy should I merge this one for now or do you want to take a closer look?

**Comment by mvc1006** [comment]
thanks




| |
陈鹏菲
|
|
邮箱：mvc1006@163.com
|

签名由 网易邮箱大师 定制

On 06/20/2019 10:03, Mr.doob wrote:

@donmccurdy should I merge this one for now or do you want to take a closer look?

**Review by donmccurdy — APPROVED** [review]
Good to merge. 👍

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**